[ONNX][TypePromo] Materialize type promotion rules

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104491
* #104229
* #104064
* __->__ #104063

This PR adds the materialized type promotion rule set and the type promotion rule infrastructure
that will be consumed by ONNX exporter. The script that generates the rule set is included, and
a local unittest is added to check the validity of the materialized rule set.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR

BowenBao committed:
[ONNX][TypePromo] Materialize type promotion rules

[ghstack-poisoned]

BowenBao committed:
rebase on "[ONNX][TypePromo] Materialize type promotion rules"




[ghstack-poisoned]

BowenBao committed:
Update rule set; Support node re-dispatch. Support tuple/list fx arg on "[ONNX][TypePromo] Materialize type promotion rules"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][TypePromo] Materialize type promotion rules"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][TypePromo] Materialize type promotion rules"




[ghstack-poisoned]

thiagocrepaldi reviewed (commented):
Consider investing on the PR description. This is a big piece of code that will be revisited in the feature by future us or new hires. Go through the problem being solved, the approach, limitations and capabilities, if possible.

BowenBao committed:
Update rule docs and comments; Relax dispatcher check when producing multiple ops on "[ONNX][TypePromo] Materialize type promotion rules"


This PR adds the materialized type promotion rule set and the type promotion rule infrastructure
that will be consumed by ONNX exporter. The script that generates the rule set is included, and
a local unittest is added to check the validity of the materialized rule set.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -0,0 +1,922 @@
+# NOTE: BELOW TABLE IS GENERATED FROM `TypePromotionRuleSetGenerator.generate_from_torch_refs`.
+# DO NOT EDIT MANUALLY !!!
+# For missing rules or discrepancies, please
+# 1. Run `pytest test/onnx/test_fx_type_promotion.py` to validate if the generated rule set is current.
```
How often should we update on this?

BowenBao replied:
Update only when needed. For regressions we have op tests to cover.

titaiwangms replied:
In that case, we have to hurry up on the op_test coverage on converter side.

BowenBao replied:
If you want feel free to attempt https://github.com/microsoft/onnxscript/issues/795, https://github.com/microsoft/onnxscript/issues/797. I have some ideas we can discuss.

titaiwangms replied:
Sure, I am interested in updating test_fx_op_consistency.py. I will join you guys after the applying param_schema/opschema on dispatcher stuff.

titaiwangms commented on test/onnx/test_fx_type_promotion.py:
```diff
@@ -0,0 +1,27 @@
+class TestGeneratedTypePromotionRuleSet(TestCase):
+    @pytorch_test_common.skip_in_ci(
+        "Reduce noise in CI. "
+        "The test serves as a tool to validate if the generated rule set is current. "
```
When should we run this?

BowenBao replied:
When there is type related issues and someone follows `Note add type promotion rule` to check if generated set is current.

titaiwangms commented on test/onnx/test_fx_type_promotion.py:
```diff
@@ -0,0 +1,27 @@
+
+    def test_initialize_type_promotion_table_from_generated_set_success(self):
+        type_promotion.TypePromotionTable()
+        self.assertTrue(True)
```
It's kind of confusing. Could you add NOTE

BowenBao replied:
I can remove L23. the test name describes the expected outcome, but I'll remove `from_generated_set` from the line as it is internal details.

justinchuby commented on test/onnx/test_fx_type_promotion.py:
```diff
@@ -0,0 +1,27 @@
+
+import pytorch_test_common
+from torch.onnx._internal.fx.passes import type_promotion
+from torch.testing._internal.common_utils import run_tests, TestCase
```
Reminder to import modules only

justinchuby commented on test/onnx/test_fx_type_promotion.py:
```diff
@@ -0,0 +1,27 @@
+from torch.testing._internal.common_utils import run_tests, TestCase
+
+
+class TestGeneratedTypePromotionRuleSet(TestCase):
```
```suggestion
class TestGeneratedTypePromotionRuleSet(common_utils.TestCase):
```

justinchuby commented on test/onnx/test_fx_type_promotion.py:
```diff
@@ -0,0 +1,27 @@
+
+
+if __name__ == "__main__":
+    run_tests()
```
```suggestion
    common_utils.run_tests()
```

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -0,0 +1,922 @@
+# NOTE: BELOW TABLE IS GENERATED FROM `TypePromotionRuleSetGenerator.generate_from_torch_refs`.
+# DO NOT EDIT MANUALLY !!!
+# For missing rules or discrepancies, please
+# 1. Run `pytest test/onnx/test_fx_type_promotion.py` to validate if the generated rule set is current.
```
Should this be a "test" or a "utility"?

BowenBao replied:
Both are fine, I feel test is more convenient to use given the current code.
When we use proper codegen we can move it to a better location, or create a separate tool.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -0,0 +1,922 @@
+        )
+
+
+# NOTE: BELOW TABLE IS GENERATED FROM `TypePromotionRuleSetGenerator.generate_from_torch_refs`.
```
Maybe we should have an exclusive file with the auto generated content? That would make future update easier, maybe even automated without having to parse stuff like "# DO NOT EDIT MANUALLY !!!"

BowenBao replied:
Created https://github.com/pytorch/pytorch/issues/104452

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -0,0 +1,922 @@
+
+
+class TypePromotionRule:
+    """Defines how to perform type promotion for 'torch.ops.{namespace}.{op_name}'."""
```
Maybe document how to "interpret" the promotion rules`so that we others can easily check whether the rules are correct, etc

```python
    TypePromotionRule(
        "aten", "bitwise_and_", [0, 1], [], ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
    ),
    TypePromotionRule(
        "aten",
        "bitwise_left_shift",
        [0, 1],
        [],
        ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
    ),
```

thiagocrepaldi replied:
Also document how `promotion_kind` works and its different values such as `ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT, ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT, ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT, ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT`

BowenBao replied:
Not quite following the interpret part, could you explain?

The docstring is added in the next PR in stack, a mistake when splitting the commits but i'll just leave it as is

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -0,0 +1,922 @@
+    ),
+}
+
+# Manually curated extra type promotion rules.
```
Try to explain why the manual rules are needed, including what is the missing piece on the auto generated part.
Could this be automated if any fix in pytorch?

BowenBao replied:
It is explained more in `[Note] Update type promotion rule` right above. I'll call it out here too.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -0,0 +1,922 @@
+
+
+class TypePromotionRuleSetGenerator:
+    """Hackly distilling info from reference ops decorated with type promotion rule."""
```
show an example of what the decorator being parsed and what info we care about from it and what the generated rule would look like

BowenBao replied:
Good point, I will mention the high level goal we want to achieve.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -0,0 +1,922 @@
+
+
+class TypePromotionTable:
+    """Type promotion table for torch.ops."""
```
how this table look like and how it will be used?

BowenBao replied:
It only has two methods `get_rule` and `add_rule`, both with docstring and type annotation so I think it's self explanatory?

thiagocrepaldi reviewed (commented):
The code look good. I would advise adding more docstrings going through some examples

BowenBao committed:
Add comments & cleanup refactor on "[ONNX][TypePromo] Materialize type promotion rules"


This PR adds the materialized type promotion rule set and the type promotion rule infrastructure
that will be consumed by ONNX exporter. The script that generates the rule set is included, and
a local unittest is added to check the validity of the materialized rule set.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR

[ghstack-poisoned]

BowenBao commented:
Going ahead and merge this. Feel free to leave new comments and I will follow up in PRs.

BowenBao commented:
@pytorchbot merge
